renderer.js

```diff
@@ -1,160 +1,188 @@
-			stepDivs[idx].appendChild(tt);
-		});
-		textValues.forEach((tv, idx) => {
-			tv.className = 'text-value';
-			stepDivs[idx].appendChild(tv);
-		});
-		canvases.forEach((canvas, idx) => {
-			canvas.width = 50;
-			canvas.height = 250;
-			stepDivs[idx].appendChild(canvas);
-		});
-		stepDivs.forEach((s) => {
-			s.className = 'step';
-			container.appendChild(s);
-		});
-	}
-
-	textThresholds.forEach((tt, idx) => (tt.innerText = thresholds[idx]));
-	textValues.forEach((tv, idx) => (tv.innerText = stepReadings[idx]));
-	canvases.forEach((canvas, idx) =>
-		drawCanvas(canvas, stepReadings[idx], thresholds[idx])
-	);
+  const s = data.toString();
+  if (s.startsWith("v")) {
+    readings = Array.from(s.match(/ \d+/g)).map((i) =>
+      Number.parseInt(i.trim())
+    );
+    updateDisplay();
+  } else if (s.startsWith("t")) {
+    thresholds = Array.from(s.match(/ \d+/g)).map((i) =>
+      Number.parseInt(i.trim())
+    );
+    updateDisplay();
+  }
+};
+
+const updateDisplay = () => {
+  if (nSteps != thresholds.length) {
+    clear();
+    nSteps = thresholds.length;
+    const container = document.querySelector("#container");
+    stepDivElements = thresholds.map(() => document.createElement("div"));
+    canvasElements = thresholds.map(() => document.createElement("canvas"));
+    textThresholdElements = thresholds.map(() =>
+      document.createElement("span")
+    );
+    textValueElements = thresholds.map(() => document.createElement("span"));
+    textThresholdElements.forEach((e, idx) => {
+      e.className = "text-threshold";
+      e.innerText = thresholds[idx];
+      stepDivElements[idx].appendChild(e);
+    });
+    textValueElements.forEach((e, idx) => {
+      e.className = "text-value";
+      stepDivElements[idx].appendChild(e);
+    });
+    canvasElements.forEach((e, idx) => {
+      e.style.width = `calc(${100 / canvasElements.length}vw - 12px)`;
+      e.style.height = `calc(100vh - 75px)`;
+      stepDivElements[idx].appendChild(e);
+    });
+    stepDivElements.forEach((e) => {
+      e.className = "step";
+      container.appendChild(e);
+    });
+  }
+
+  textThresholdElements.forEach((tt, idx) => (tt.innerText = thresholds[idx]));
+  textValueElements.forEach((tv, idx) => (tv.innerText = readings[idx]));
+  canvasElements.forEach((canvas, idx) =>
+    drawCanvas(canvas, readings[idx], thresholds[idx])
+  );
 };
 
 const drawCanvas = (canvas, value, threshold) => {
-	const ctx = canvas.getContext('2d');
-	ctx.clearRect(0, 0, canvas.width, canvas.height);
-	ctx.fillStyle = value > threshold ? '#BF616A' : '#5E81AC';
-	const normalizedValue = value / maxValue;
-	const top = normalizedValue * canvas.height;
-	ctx.fillRect(0, canvas.height - top, canvas.width, top);
-
-	ctx.fillStyle = '#ECEFF4';
-	const normalizedThreshold = threshold / maxValue;
-	ctx.fillRect(
-		0,
-		canvas.height - normalizedThreshold * canvas.height - 3,
-		canvas.width,
-		3
-	);
+  let canvasWidth = canvas.clientWidth;
```

Comment:
Since the canvas sizes are dynamic now, we can fetch them with this instead of using their default values which won't ever scale.